File: typescript/LWTSD/ResourceTypes/ResourceInteger.ts
var xml = require('../../node-xmpp/packages/xml');

import { LWTSD } from '../index';
import { SimplifiedType } from '../SimplifiedType';
import { ResourceDescription } from '../ResourceDescription'
import { ResourceRestriction } from '../ResourceRestriction';
import { Resource } from './Resource'

// This should have some proper integer checks..
export class ResourceInteger extends Resource
{
    public readonly HistoricalValues :  Array< [Date, number] > = new  Array< [Date, number] > ();

    InnerValue : number;

    public GetValue() : number
    {
        return this.InnerValue;
    }

    public SetValue(Value : number) : void
    {
        if(this.InnerMinExclusive && Value < (this.InnerMinExclusive as Number))
            throw new Error("Value not within limits");
        if(this.InnerMaxExclusive && Value > (this.InnerMaxExclusive as Number))
            throw new Error("Value not within limits");

        this.InnerValue = Value;
        this.ModifiedAt = new Date();
    }

    public constructor(Element? : any)
    {
        super(Element);

        let points = Element.getChildren("point");
        for (let i = 0; i < points.length; i++)
        {
            let it = points[i];
            let TimeStamp = new Date(it.attrs.timestamp);
            let PointValue : number = parseInt(it.text());
            this.HistoricalValues.push( [TimeStamp, PointValue] );
            if (TimeStamp > this.ModifiedAt)
            {
                this.ModifiedAt = TimeStamp;
                this.InnerValue = PointValue;
            }
        }
    }

    public GetPoint() : any
    {
        let RVal = xml("point", { /*"xmlns": LWTSD.Namespace */}, this.GetValue());
        RVal.attrs.timestamp = this.ModifiedAt;
        return RVal;
    }

    public GetWriteValue() : string
    {
        return this.GetValue().toString();
    }

    public GetDescription() : ResourceDescription
    {
        return super.GetDescriptionFromSimpleType(SimplifiedType.Float);
    }

    public LoadFromWrite(WriteElement : any) : void
    {
        this.SetValue( WriteElement.text() );
    }
}

File: NET/LWTSD/ResourceTypes/ResourceInteger.cs
﻿using System;
using System.Xml;
using System.Xml.Linq;
using System.Collections.Generic;

namespace LWTSD.ResourceTypes
{
	public class ResourceInteger : Resource
	{
		public readonly List<Tuple<DateTime, int>> HistoricalValues = null;

		int InnerValue;
		public int Value
		{
			get
			{
				return InnerValue;
			}

			set
			{
				if (MinExclusive != null)
					if (value < MinExclusive)
						throw new InvalidOperationException("Value not within limits");
				if (MaxExclusive != null)
					if (value > MaxExclusive)
						throw new InvalidOperationException("Value not within limits");

				InnerValue = value;

				ModifiedAt = DateTime.UtcNow;
			}
		}

		public int? MinExclusive
		{
			get
			{
				if (InnerMinExclusive == null)
					return null;

				return (Int32) InnerMinExclusive;
			}

			set
			{
				InnerMinExclusive = value;
			}
		}
		public int? MaxExclusive
		{
			get
			{
				if (InnerMaxExclusive == null)
					return null;

				return (Int32) InnerMaxExclusive;
			}

			set
			{
				InnerMaxExclusive = value;
			}
		}

		public override XElement GetPoint()
		{
			XElement Point = base.GetPoint();
			Point.Value = XmlConvert.ToString(Value);
			return Point;
		}

		public override ResourceDescription GetDescription()
		{
			return base.GetDescription(SimplifiedType.Integer);
		}

		public override void LoadFromWrite(XElement WriteElement)
		{
			this.Value = XmlConvert.ToInt32(WriteElement.Value);
		}

		public override string GetWriteValue()
		{
			return Value.ToString();
		}

		public ResourceInteger(XElement Element)
			: base(Element)
		{
			ModifiedAt = DateTime.MinValue;
			foreach (var it in Element.Elements(LWTSD.Namespace + "point"))
			{
				if (HistoricalValues == null)
					HistoricalValues = new List<Tuple<DateTime, int>>();
				DateTime TimeStamp = XmlConvert.ToDateTime(it.Attribute("timestamp").Value,
														   XmlDateTimeSerializationMode.Utc);
				int PointValue = XmlConvert.ToInt32( it.Value );
				HistoricalValues.Add(new Tuple<DateTime, int>(TimeStamp, PointValue));

				if (TimeStamp > ModifiedAt)
				{
					ModifiedAt = TimeStamp;
					InnerValue = PointValue;
				}

			}
		}

		public ResourceInteger()
		{
		}

	}
}
